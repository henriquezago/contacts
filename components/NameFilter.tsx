import { Input } from "@nextui-org/react";
import { useEffect, useState } from "react";

import useDebounce from "../hooks/useDebounce";

type NameFilterProps = {
  filter: string;
  onChange: (name: string) => void;
}

export const DEBOUNCE_DELAY = 300;

function NameFilter({ filter, onChange }: NameFilterProps) {
  const [name, setName] = useState<string>(filter);

  const debouncedName = useDebounce(name, DEBOUNCE_DELAY);

  useEffect(() => {
    onChange(debouncedName);
  }, [debouncedName]);

  return (
    <Input
      id="name-filter"
      type="search"
      aria-label="Filter contacts by name"
      placeholder="Filter by name"
      value={name}
      size="lg"
      onChange={e => setName(e.target.value)}
    />
  );
}

export default NameFilter;
