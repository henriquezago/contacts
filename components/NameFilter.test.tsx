import { expect } from '@jest/globals';
import { act, fireEvent, render } from "@testing-library/react";

import NameFilter, { DEBOUNCE_DELAY } from "./NameFilter";

describe('NameFilter', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('should render', () => {
    const { container } = render(<NameFilter onChange={jest.fn()} filter="" />);
    expect(container).toMatchSnapshot();
  });

  it('should call onChange after delay', () => {
    const onChange = jest.fn();
    const { container } = render(<NameFilter onChange={onChange} filter="" />);

    act(() => {
      const inputElement = container.querySelector("#name-filter");
      fireEvent.change(inputElement, { target: { value: 'test' } });
      
      jest.advanceTimersByTime(DEBOUNCE_DELAY + 1);
    });


    expect(onChange).toHaveBeenCalledWith('test');
  });
});