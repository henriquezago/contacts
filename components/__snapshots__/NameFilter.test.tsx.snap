// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameFilter should render 1`] = `
<div>
  <div
    class="nextui-c-iwmZdB nextui-c-iwmZdB-jTfqsF-helperColor-default nextui-c-iwmZdB-dpkacN-borderWeight-normal nextui-c-iwmZdB-bkiCag-size-lg nextui-c-iwmZdB-iiIdOsz-css nextui-input-main-container nextui-input-main-container--normal"
    data-state="normal"
  >
    <div
      class="nextui-c-boZsAS nextui-c-boZsAS-fjttfO-isTextarea-false nextui-c-boZsAS-iNKybL-animated-true nextui-input-container nextui-input-container--normal nextui-input-container--input"
    >
      <label
        class="nextui-c-eXOOPO nextui-c-eXOOPO-iNKybL-shadow-true nextui-c-eXOOPO-ccrOPm-underlined-false nextui-c-eXOOPO-lkBNdM-isTextarea-false nextui-input-wrapper nextui-input-wrapper--normal nextui-input-wrapper--shadow"
      >
        <span
          class="nextui-c-PJLV nextui-c-iFHsrU nextui-input-placeholder"
        >
          Filter by name
        </span>
        <input
          aria-label="Filter contacts by name"
          aria-multiline="false"
          aria-placeholder="Filter by name"
          aria-readonly="false"
          aria-required="false"
          autocomplete="off"
          class="nextui-c-jeuecp nextui-input"
          data-state="normal"
          id="name-filter"
          placeholder="Filter by name"
          type="search"
          value=""
        />
      </label>
    </div>
    <div
      class="nextui-c-gJvMas nextui-input-helper-text-container"
    />
  </div>
</div>
`;
